### PR TITLE
Extend AllocMonitor

### PR DESCRIPTION
**Pull request by Dr15Jones**
#### PR description:

- added overriding of additional allocation methods which are used by 3rd party libraries used by CMSSW
- added PeriodicAllocMonitor service

#### PR validation:

All package related unit tests pass. The work was first done in a local CMSSW_13_2_3 work area to find problems in HI workflows and then moved to master.

### CODE REVIEW

**Comment by makortel** [comment]
@cmsbuild, please test

File: PerfTools/AllocMonitor/README.md [hunk 1]
```diff
@@ -77,3 +77,25 @@ bin is a power of 2 larger than the previous. The histograms made are
 - Amount of bytes actually returned by all deallocation calls
 
 This service is multi-thread safe. Note that when run multi-threaded the maximum reported value will vary from job to job.
+
+### PeriodicAllocMonitor
+This service registers a monitor when the service is created (after python parsing is finished but before any modules
+have been loaded into cmsRun) and prints its accumulated information to the specified file at specified intervals. Both
+the file name and  interval are specfied by setting parameters of the service in the configuration. The parameters are
```
Comment (makortel) [inline PerfTools/AllocMonitor/README.md:84]:
```suggestion
the file name and  interval are specified by setting parameters of the service in the configuration. The parameters are
```

File: PerfTools/AllocMonitor/plugins/PeriodicAllocMonitor.cc [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+        fs << std::chrono::duration_cast<std::chrono::milliseconds>(now - start).count() << ", " << rStarted << ", "
+           << lStarted << ", " << eStarted << ", " << eFinished << ", " << report.requested_ << ", "
+           << report.maxActual_ << ", " << report.presentActual_ << ", " << report.maxSingleRequested_ << ", "
+           << report.nAllocations_ << ", " << report.nDeallocations_ << std::endl;
+        std::this_thread::sleep_for(std::chrono::milliseconds(interval));
+      }
+    });
+
+    //iAR.postBeginJobSignal_.connect([adaptor]() { adaptor->start(); });
+    iAR.preEventSignal_.connect([this](auto const&) { nEventsStarted_.fetch_add(1, std::memory_order_acq_rel); });
```
Comment (makortel) [inline PerfTools/AllocMonitor/plugins/PeriodicAllocMonitor.cc:122]: Could these use the `iAR.watchPreEvent()` interface to register the callbacks? (not a big deal, and I see `EventProcessingAllocMonitor` already slipped through)

File: PerfTools/AllocMonitor/test/test_proxies.cc [hunk 3]
```diff
@@ -182,17 +182,50 @@ int main() {
...
       free(p);
       if (total != 0) {
         std::cout << "aligned_alloc request not cleaned up";
         exit(1);
       }
+
+      p = memalign(256, 24);
+      if (requested != 24) {
+        auto r = requested;
+        std::cout << "aligned_alloc request size wrong, got " << r << " expected " << 24;
```
Comment (makortel) [inline PerfTools/AllocMonitor/test/test_proxies.cc:200]: Should this be
```suggestion
        std::cout << "memalign request size wrong, got " << r << " expected " << 24;
```
?

File: PerfTools/AllocMonitor/test/test_proxies.cc [hunk 4]
```diff
@@ -182,17 +182,50 @@ int main() {
...
+        exit(1);
+      }
+      free(p);
+      if (total != 0) {
+        std::cout << "aligned_alloc request not cleaned up";
```
Comment (makortel) [inline PerfTools/AllocMonitor/test/test_proxies.cc:205]:
```suggestion
        std::cout << "memalign request not cleaned up";
```
?

File: PerfTools/AllocMonitor/README.md (hunk 1 above)
Comment (Dr15Jones) [reply PerfTools/AllocMonitor/README.md]: done

File: PerfTools/AllocMonitor/test/test_proxies.cc (hunk 3 above)
Comment (Dr15Jones) [reply PerfTools/AllocMonitor/test/test_proxies.cc]: done

File: PerfTools/AllocMonitor/plugins/PeriodicAllocMonitor.cc (hunk 2 above)
Comment (Dr15Jones) [reply PerfTools/AllocMonitor/plugins/PeriodicAllocMonitor.cc]: done

**Comment by Dr15Jones** [comment]
please test

**Comment by makortel** [comment]
+core

**Comment by rappoccio** [comment]
+1

**Pull request merged**